docs/proposals/volume-provisioning.md

```diff
@@ -0,0 +1,509 @@
+
+2.  A new API group called `storage` be created to hold the a `StorageClass`
+    API resource. Storage configuration is represented as `LabelSelector`,
+    specifying parameters of volumes belonging to this class.
+
+3.  `PersistentVolume.Spec.Class` attribute is added to volumes. This attribute
+    is optional and specifies which `StorageClass` instance represents
+    storage characteristics of a particular PV.
+
+    During alpha and beta development, `Class` is just an annotation and not
+    actual attribute.
+
+4.  `PersistentVolume` instances should be labelled with configuration
+    parameters that were used for their creation, such as availability zone or
+    storage type (SSD/HDD). This should be done in an admission controller.
+
+5.  `PersistentVolumeClaim.Spec.Class` attribute is added to claims. This
+    attribute specifies that only a volume with equal
+    `PersistentVolume.Spec.Class` value can satisfy a claim.
+
+    During alpha and beta development, `Class` is just an annotation and not
+    actual attribute.
+
+6.  When requesting a volume with certain configuration,
+    `PersistentVolumeClaim.Spec.Selector` should be set in a way that it matches
+    only volumes with this configuration (see point 3.)
+    When both `PersistentVolumeClaim.Spec.Selector` and
+    `PersistentVolumeClaim.Spec.Class` is set, resulting volume must match both
+    selectors in the claim and in the storage class.
+
+7.  The persistent volume controller be modified to dynamically provision
+    volumes using storage classes and claim's `Selector` to create a volume that
+    has properties requested by both `PersistentVolumeClaim.Spec.Class` and
+    `PersistentVolumeClaim.Spec.Selector`.
+
+8.  The existing provisioner plugin implementations be modified to allow
+    parameterization as appropriate via storage classes and claim selectors.
+
+9.  The existing alpha dynamic provisioning feature be phased out in the
+    next release.
+
+### Controller workflow for provisioning volumes
+
+1.  When a new claim is submitted, the controller attempts to find an existing
+    volume that will fulfill the claim.
+
+    1. If the claim has non-empty `claim.Spec.Class`, only PVs with the same
+        `pv.Spec.Class` are considered.
+
+    2. If the claim has empty `claim.Spec.Class`, all existing PVs are
+        considered.
+
+    All "considered" volumes are evaluated with `claim.Spec.Selector` and the
+    smallest matching one is then bound to the claim.
+
+2.  If no volume is found for the claim and `claim.Spec.Class` is not set or is
+    empty string, the controller will use a default StorageClass specified on
+    kube-controller-manager command line. If no default StorageClass is
+    configured, dynamic provisioning is disabled and the controller goes back to
+    step 1. and periodically retries finding a matching volume or storage class
+    again until a match is found. The claim is `Pending` during this period.
```

Comment:
I think I'd rather it just failed/go to step 1 and we work on how to handle defaulting in a follow up. If we want to force a storage class i'd rather see it as an admission controller instead of an command line flag on the kube-controller-manager